File: algorithms/cpp/k-diffPairsinanArray/k-diffPairsinanArray.cpp
// Source: https://leetcode.com/problems/k-diff-pairs-in-an-array/
// Author: Miao Zhang
// Date:   2021-02-20

class Solution {
public:
    int findPairs(vector<int>& nums, int k) {
        unordered_map<int, int> dic;
        for (auto n: nums) dic[n]++;
        int res = 0;
        for (auto &it: dic) {
            if (k > 0 && dic.count(it.first + k)) res++;
            if (k == 0 && it.second > 1) res++;
        }
        return res;
    }
};

File: algorithms/python/k-diffPairsinanArray/k-diffPairsinanArray.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-

# Source: https://leetcode.com/problems/k-diff-pairs-in-an-array/
# Author: Miao Zhang
# Date:   2021-02-20

class Solution:
    def findPairs(self, nums: List[int], k: int) -> int:
        counter = collections.Counter(nums)
        res = 0
        for n in set(nums):
            if k > 0 and n + k in counter:
                res += 1
            if k == 0 and counter[n] > 1:
                res += 1
        return res
